File: stack/150_evaluate_reverse_polish_notation.go
// O(n) time | O(n) space - where n is tokens.length
func evalRPN(tokens []string) int {
    stack := []int{}
    
    ops := map[string]func(int, int) int{
        "+": func(a, b int) int { return a + b },
        "-": func(a, b int) int { return a - b },
        "*": func(a, b int) int { return a * b },
        "/": func(a, b int) int { return int(a / b) },
    }

    for _, token := range tokens {
        if op, ok := ops[token]; ok {
            a, b := stack[len(stack) - 2], stack[len(stack) - 1]
            stack = stack[:len(stack) - 2]
            stack = append(stack, op(a, b))
        } else {
            num, _ := strconv.Atoi(token)
            stack = append(stack, num)
        }
    }

    return stack[0]
}

File: stack/150_evaluate_reverse_polish_notation.py
class Solution:
    # O(n) time | O(n) space - where n is tokens.length
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = {
            "+" : lambda a, b: a + b,
            "-" : lambda a, b: a - b,
            "*" : lambda a, b: a * b,
            "/" : lambda a, b: int(a / b)
        }
        for token in tokens:
            if token in ops:
                b, a = stack.pop(), stack.pop()
                stack.append(ops[token](a, b))
            else:
                stack.append(int(token))
        return stack[0]

